### PR TITLE
Add JObjectAdapter to support JSON Patch for JObject properties

### PR DESCRIPTION
**Pull request by sfmskywalker**
Added `JObjectAdapter` to support JSON Patch for JObject properties
 - Updated AdapterFactory to test against object type: if it's `JObject`, return new `JObjectAdapter`

Addresses #2429

### CODE REVIEW

**Comment by mkArtakMSFT** [comment]
Thanks for your effort, @sfmskywalker.
@rynowak, @pranavkm can you please review this? Thanks!

**Comment by sfmskywalker** [comment]
No problem! Let me know if anything looks off. I just pushed another update in hopes the final 4 checks will also pass.

**Comment by rynowak** [comment]
The code changes here look good. Can you add some tests that use this? 

There are some examples of things that don't work today in the original bug.

**Comment by sfmskywalker** [comment]
Sure, I will add some unit tests using this.

**Comment by sfmskywalker** [comment]
Just a quick update - I successfully tested all the cases mentioned in the original bug (#2429) with a console application. I have yet to figure out how to get this repository to compile before I can start adding unit tests.

**Comment by sfmskywalker** [comment]
Yeah apologies for the delay. I’m gonna give it another try over the weekend.

**Comment by sfmskywalker** [comment]
I keep getting the following error when I try to build using build.cmd:

```shell
PS C:\Projects\sfmskywalker\AspNetCore> .\build.cmd
WARNING: No default group of projects was specified, so building the 'managed' and its dependent subsets of projects. Run `build.cmd -help` for more details.
Building of C# project is enabled and has dependencies on NodeJS projects. Building of NodeJS projects is enabled since node is detected in C:\Program Files.
Detected JDK in C:\Projects\sfmskywalker\AspNetCore\.tools\jdk\win-x64\ (via local repo convention)
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-alpha1-013788/dotnet-sdk-5.0.100-alpha1-013788-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-alpha1-013788/dotnet-sdk-5.0.100-alpha1-013788-win-x64.zip
dotnet-install: Adding to current process PATH: "C:\Projects\sfmskywalker\AspNetCore\.dotnet\". Note: This change will not be visible if PowerShell was run as a child process.
dotnet-install: Installation finished
C:\Users\SipkeSchoorstra\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19411.1\tools\Tools.proj : error : Unable to locate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the installed version.
C:\Users\SipkeSchoorstra\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19411.1\tools\Tools.proj : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
Build failed.
build.cmd completed
```

My fork is up to date with today's master branch. Looking at `global.json`, I see a reference being made to SDK version `5.0.100-alpha1-013788`. I could not find an installer for that version anywhere. 

I must be doing something wrong.
Any pointers?

**Comment by sfmskywalker** [comment]
It looks like I'm not the only one having an issue with building the source. I added my comments and questions here: https://github.com/aspnet/AspNetCore/issues/12913#issuecomment-526605903

**Comment by sfmskywalker** [comment]
Hang on, I only now notice the following line:

```
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-alpha1-013788/dotnet-sdk-5.0.100-alpha1-013788-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-alpha1-013788/dotnet-sdk-5.0.100-alpha1-013788-win-x64.zip
```

Clearly it's downloading the latest SDK the build is failing to find.
I'll dig some more.

**Comment by sfmskywalker** [comment]
Trying the suggestion here: https://github.com/aspnet/AspNetCore/issues/11566

**Comment by sfmskywalker** [comment]
I got the build working by following these steps: https://github.com/aspnet/AspNetCore/issues/12913#issuecomment-526823482

**Comment by sfmskywalker** [comment]
So, in order to add tests, I wanted to add a `JsonPatch.sln` file and add both the library project and the existing test project to it. Since there's now .sln file yet, I followed the instructions and tried the following:

*Path*
`(AspNetCore) PS C:\Projects\sfmskywalker\AspNetCore>`

```ps
cd C:\Projects\sfmskywalker\AspNetCore\src\Features\JsonPatch
dotnet new sln -n JsonPatch
dotnet sln add ./src/Microsoft.AspNetCore.JsonPatch.csproj
```

So far, so good - the solution file was created and the .NET standard 2.0 `Microsoft.AspNetCore.JsonPatch.csproj` library got added.

Now I want to add the test project:

```ps
dotnet sln add ./test/Microsoft.AspNetCore.JsonPatch.Tests.csproj
```

Which didn't work:

`Invalid project `C:\Projects\sfmskywalker\AspNetCore\src\Features\JsonPatch\test\Microsoft.AspNetCore.JsonPatch.Tests.csproj`. The SDK 'Microsoft.NET.Sdk' specified could not be found.`

But when I do `dotnet --list-sdks`, I see:

```
3.0.100-preview7-012821 [C:\Projects\sfmskywalker\AspNetCore\.dotnet\sdk]
5.0.100-alpha1-013788 [C:\Projects\sfmskywalker\AspNetCore\.dotnet\sdk]
```

And when I do `dotnet --version`, I get:

```
5.0.100-alpha1-013788
```

So I thought maybe the `DefaultNetCoreTargetFramework` variable isn't set, so Itried updating the .csproj with a `TargetFrameworks` of `netcoreapp5.0`:

```xml
<PropertyGroup>
<TargetFrameworks>netcoreapp5.0</TargetFrameworks>
</PropertyGroup>
```

But that didn't make any difference.

So despite the fact that `dotnet` is set to use the latest preview SDK (5.0.100-alpha1-013788), it can't seem to understand the specified target framework in `Microsoft.AspNetCore.JsonPatch.Tests.csproj`

And yet, running .\build on existing .sln files, such as MVC.sln, works fine (except for the issue with some long-winded file path not being able to be copied).

Obviously I'm missing something, but I don't yet see what that is.

**Comment by rynowak** [comment]
@sfmskywalker - could you update your branch with those changes (adding sln file) and I'll give it a try?

> Invalid project C:\Projects\sfmskywalker\AspNetCore\src\Features\JsonPatch\test\Microsoft.AspNetCore.JsonPatch.Tests.csproj`. The SDK 'Microsoft.NET.Sdk' specified could not be found.

What are the repro steps (exactly) for this error so I can see it and investigate?

**Comment by SipkeSchoorstraRTL** [comment]
@rynowak Great, thank you. I just updated my branch by adding the .SLN file. Steps to reproduce the error:

1. Open a command prompt / PowerShell in the root directory of the project.
2. Run `. .\activate.ps1` 
3. `cd src\Features\JsonPatch`
4. `dotnet sln add ./test/Microsoft.AspNetCore.JsonPatch.Tests.csproj`

When I do the above, I receive the following error:

`Invalid project `C:\Projects\sfmskywalker\AspNetCore\src\Features\JsonPatch\test\Microsoft.AspNetCore.JsonPatch.Tests.csproj`. The SDK 'Microsoft.NET.Sdk' specified could not be found.`

**Comment by sfmskywalker** [comment]
@rynowak Great, thank you. I just updated my branch by adding the .SLN file. Steps to reproduce the error:

1. Open a command prompt / PowerShell in the root directory of the project.
2. Run `. .\activate.ps1` 
3. `cd src\Features\JsonPatch`
4. `dotnet sln add ./test/Microsoft.AspNetCore.JsonPatch.Tests.csproj`

When I do the above, I receive the following error:

`Invalid project C:\Projects\sfmskywalker\AspNetCore\src\Features\JsonPatch\test\Microsoft.AspNetCore.JsonPatch.Tests.csproj. The SDK 'Microsoft.NET.Sdk' specified could not be found.`

**Comment by mkArtakMSFT** [comment]
@rynowak is this ready to be merged?

**Comment by rynowak** [comment]
I tried this out and I didn't have any issues. There's a branch here with an update to the solution file: https://github.com/aspnet/AspNetCore/tree/issue/jsonpatch-jobject

If you're still having problems make sure you run `restore.cmd` in the repo root.

**Comment by sfmskywalker** [comment]
I’ll give it a go. Thanks!

**Comment by sfmskywalker** [comment]
@rynowak The solution works and I'm now able to actually build & run tests 😄 . Expect an update shortly.

**Comment by sfmskywalker** [comment]
Tests have been added for all patch operations in combination with types containing `JObject` properties ✔️

File: src/Features/JsonPatch/test/JsonPatchDocumentJObjectTest.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        [Fact]
+        public void ApplyTo_ExpandoObject_AddAndReplace()
+        {
+            // Arrange
+            var list = new List<string>
+            {
+                "[{ 'op': 'add', 'path': '/customData', 'value': {'bar': 'baz'}}, { 'op': 'replace', 'path': '/customData/bar', 'value': 'bazz' }]"
+            };
+
+            dynamic model = new ExpandoObject();
```
Comment (rynowak) [inline src/Features/JsonPatch/test/JsonPatchDocumentJObjectTest.cs]: Maybe I'm missing something - why is `ExpandoObject` relevant here?

Comment (sfmskywalker) [reply src/Features/JsonPatch/test/JsonPatchDocumentJObjectTest.cs]: My mistake - I misinterpreted the issue from the OP and thought there was an issue with ExpandoObject in combination with JObject. But after re-reading, it seems that he was simply using ExpandoObject to work around the issue due to the lack of an JObjectAdapter.

I’ll get this test removed 👍🏻

Comment (rynowak) [reply src/Features/JsonPatch/test/JsonPatchDocumentJObjectTest.cs]: OK great.

Comment (rynowak) [reply src/Features/JsonPatch/test/JsonPatchDocumentJObjectTest.cs]: Send a ping when this is done and I'll see if I can merge it. We've been having some CI issues today which is why the builds are red.

Comment (sfmskywalker) [reply src/Features/JsonPatch/test/JsonPatchDocumentJObjectTest.cs]: @rynowak All set.

**Review by rynowak — APPROVED** [review]

**Pull request merged**